storage/innobase/include/data0type.ic

```diff
@@ -625,6 +631,9 @@ dtype_get_sql_null_size(
 	const dtype_t*	type,	/*!< in: type */
 	ulint		comp)	/*!< in: nonzero=ROW_FORMAT=COMPACT  */
 {
+	if (type->mtype == DATA_CHAR || type->mtype == DATA_INT) {
+		return(type->len);
+	}
 	return(dtype_get_fixed_size_low(type->mtype, type->prtype, type->len,
 					type->mbminlen, type->mbmaxlen, comp));
```

Comment:
For any value we reserve space of `len`, i.e. 0 and NULL will reserve maximum possible space. This can make sense if we remember minimum used space like we do now for Case 4 with `unsigned_len`. Currently it has not much sense, except making tests happy. I was too doubtful whether this needed or not.

Reply:
And there is no test coverage for reserved space in MDEV-15563...